tests: adapt QVariant compare and QgsFlagKeysToValue tests to qt6

benoitdm-oslandia opened the pull request:
This PR fixes the test_core_qgis test to work with QT6.

benoitdm-oslandia committed:
fix: make qgsVariantCompare for QString the same between qt5 and qt6

qt5 return -1, 0, 1 where qt6 returns the lexical distance between the 2 strings.

rouault commented on tests/src/core/testqgis.cpp:
```diff
@@ -543,7 +548,7 @@ void TestQgis::testQgsFlagKeysToValue()
   QCOMPARE( ok, true );
   // also try with an invalid int value
-  QCOMPARE( qgsFlagKeysToValue( QString::number( -1 ), defaultValue, true, &ok ), defaultValue );
+  QCOMPARE( qgsFlagKeysToValue( QString::number( std::numeric_limits<int>::quiet_NaN() ), defaultValue, true, &ok ), defaultValue );
```
``std::numeric_limits<int>::quiet_NaN()`` evaluates to 0. Is that what you mean ?

benoitdm-oslandia replied:
No, I was sure it evaluates to a NaN value :-1:  
Here the issue is negative number are no more invalid, do you think it will break the test if I use `std::numeric_limits<double>::quiet_NaN()`?

benoitdm-oslandia replied:
I have tried with `std::numeric_limits<long>::max()` and debug it step by step to check if it takes the right code path.

rouault replied:
I'd avoid using long as it is 32 bit on Windows 64 and 64 bit on other 64-bit systems.  Isn't int enough ?

benoitdm-oslandia replied:
The issue did not come from the string to int conversion but from the call to `metaEnum.valueToKeys( intValue );` with a negative value. Qt6 returns an empty string ("") and Qt5 returns a string representation of a default ("HideReadOnly|OriginProvider|AllTypes")

nyalldawson commented on tests/src/core/testqgis.cpp:
```diff
@@ -342,8 +342,13 @@ void TestQgis::qVariantCompare_data()
   QTest::newRow( "qstringlist 6" ) << QVariant( QStringList() << QStringLiteral( "aa" ) << QStringLiteral( "xx" ) ) << QVariant( QStringList() << QStringLiteral( "aa" ) ) << false << true << 1;
   QTest::newRow( "string both empty" ) << QVariant( QString() ) << QVariant( QString() ) << false << false << 0;
+#if QT_VERSION >= QT_VERSION_CHECK( 6, 0, 0 )
+  QTest::newRow( "string" ) << QVariant( "a b c" ) << QVariant( "d e f" ) << true << false << -3;
```
I'd lean toward modifying qVariantCompare to always return -1 / +1 instead -- that'd give more predictable behavior across versions, and I can't see a strong need for non -1/0/+1 values.

benoitdm-oslandia committed:
tests: adapt QgsFlagKeysToValue to qt6

With qt6 `metaEnum.valueToKeys( -1 );` failed by returning "" instead of some default ("HideReadOnly|OriginProvider|AllTypes")

benoitdm-oslandia commented:
@nyalldawson @rouault  sorry to force push by I needed to split the commit in two

nyalldawson commented on src/core/qgis.cpp:
```diff
@@ -279,7 +279,7 @@ int qgsVariantCompare( const QVariant &lhs, const QVariant &rhs )
 
     default:
-      return QString::localeAwareCompare( lhs.toString(), rhs.toString() );
+      return std::max( -1, std::min( 1, QString::localeAwareCompare( lhs.toString(), rhs.toString() ) ) );
```
```suggestion
      return std::clamp( QString::localeAwareCompare( lhs.toString(), rhs.toString() ), -1, 1 );
```

benoitdm-oslandia commented:
@nyalldawson I will rebase and remove the draft flag

benoitdm-oslandia commented:
@nyalldawson @rouault gentle ping!
